zio-http/shared/src/main/scala/zio/http/Body.scala

```diff
@@ -568,27 +570,35 @@ object Body {
     override def asStream(implicit trace: Trace): ZStream[Any, Throwable, Byte] =
       ZStream.unwrap {
         ZIO.blocking {
-          for {
-            r <- ZIO.attempt {
+          ZIO.suspendSucceed {
+            try {
               val fs   = new FileInputStream(file)
               val size = Math.min(chunkSize.toLong, file.length()).toInt
 
-              (fs, size)
-            }
-            (fs, size) = r
-          } yield ZStream
-            .repeatZIOOption[Any, Throwable, Chunk[Byte]] {
-              for {
-                buffer <- ZIO.succeed(new Array[Byte](size))
-                len    <- ZIO.attempt(fs.read(buffer)).mapError(Some(_))
-                bytes  <-
-                  if (len > 0) ZIO.succeed(Chunk.fromArray(buffer.slice(0, len)))
-                  else ZIO.fail(None)
-              } yield bytes
+              val read: Task[Option[Chunk[Byte]]] =
+                ZIO.suspendSucceed {
+                  try {
+                    val buffer = new Array[Byte](size)
+                    val len    = fs.read(buffer)
+                    if (len > 0) Exit.succeed(Some(Chunk.fromArray(buffer.slice(0, len))))
+                    else Exit.none
+                  } catch {
+                    case e: Throwable => Exit.fail(e)
+                  }
+                }
+
+              Exit.succeed {
+                // Optimised for our needs version of `ZIO.repeatZIOChunkOption`
+                ZStream
+                  .unfoldChunkZIO(read)(_.map(_.map(_ -> read)))
```

Comment:
@kyri-petrou Funnily enough, if we were managing to find a way to write a recursive type in Scala, 

1. we wouldn't need this ugly `(_.map(_.map(_ -> read))`
2. we could write this:
```scala
              type Read = Task[Option[(Chunk[Byte], Read)]]
              val read: Read =
                ZIO.suspendSucceed {
                  try {
                    val buffer = new Array[Byte](size)
                    val len    = fs.read(buffer)
                    if (len > 0) Exit.succeed(Some(Chunk.fromArray(buffer.slice(0, len) -> read)))
                    else Exit.none
                  } catch {
                    case e: Throwable => Exit.fail(e)
                  }
                }

              Exit.succeed {
                // Optimised for our needs version of `ZIO.repeatZIOChunkOption`
                ZStream
                  .unfoldChunkZIO(read)(ZIO.identityFn)
                  .ensuring(ZIO.attempt(fs.close()).ignoreLogged)
              }
```

😍

Reply:
I can imagine, that there is a nice way to do this for Scala 3. But I think we are unable to find a nice way for Scala 2